### PR TITLE
env2mfile: Add unit test coverage for dpkg -> Meson, and fix some obvious errors

### PR DESCRIPTION
**Pull request by smcv**
* env2mfile: Convert MachineInfo into a dataclass
    
    This will make it easier to unit-test functions that work with a
    MachineInfo, by constructing the expected object in a single call.

* env2mfile: Split detect_cross_debianlike()
    
    Separating the part that runs dpkg-architecture from the part that
    interprets its results will make it easier to unit-test the latter.

* unittests: Test env2mfile's dpkg_architecture_to_machine_info
    
    This test parses several possible outputs of dpkg-architecture and
    asserts that they produce the expected MachineInfo.
    
    To avoid depending on suitable cross-tools being installed, use
    unittest.mock to override locate_path with a version that pretends that
    all of the tools we're interested in are in /usr/bin.
    Similarly, use mock environment variables to exercise what happens
    when we have those set.
    
    The test data used here exercises most variations:
    
    * big- and little-endianness
    * GNU CPU (x86_64) differing from dpkg CPU (amd64)
    * Linux, kFreeBSD and Hurd
    * special-cased architectures: x86, arm, mips64el, ppc64el
    
    expected_compilers() intentionally doesn't assume that every compiler
    is gcc (even though they all are, right now), because #13721 proposes
    adding valac which does not take a gcc suffix.

* env2mfile: Add a function for mostly-identity mappings with special cases
    
    This makes the frequent pattern of things like "CPU families are the
    same as GNU CPUs, with a few known exceptions" less verbose.

* env2mfile: Map Debian GNU/Hurd to system() -> gnu
    
    As per <https://mesonbuild.com/Reference-tables.html>, and matching what
    happens when running Meson for a native build on Debian GNU/Hurd.

* env2mfile: Don't hard-code Debian as always being Linux
    
    All official Debian release architectures use the Linux kernel, but
    unofficial ports like hurd-i386 and kfreebsd-amd64 use the Hurd and
    FreeBSD kernel, respectively.
    
    Map Linux to 'linux' and kFreeBSD ports to 'freebsd' as per the
    reference tables in Meson's documentation. For now, use the Debian
    system name such as 'hurd' for anything else (see #13740 for the
    question of whether Hurd should identify its kernel differently).

* env2mfile: Base cpu on DEB_HOST_GNU_CPU unless DEB_HOST_ARCH is special
    
    `DEB_HOST_ARCH` encodes both the CPU family and the OS, so using it to
    get the CPU type gives the wrong answer for non-Linux ports.
    
    However, `DEB_HOST_GNU_CPU` gives less detailed information about the
    CPU: it's `arm` for all 32-bit ARM CPUs, and doesn't distinguish between
    the differing baselines of `armel` (ARMv5 softfloat) and `armhf`
    (ARMv7 hardfloat).
    
    When cross-compiling for x86_64 Linux, this changes the `cpu()` from
    `amd64` to `x86_64`, which is consistent with the answer we get during
    native builds on that architecture.
    
    When cross-compiling for `ppc64el`, this changes the `cpu()` from
    `ppc64el` to `ppc64`, which is a reasonable change but is still not
    consistent with what we see during native builds (which is `ppc64le`):
    see #13741 for that.
    
    Resolves: https://github.com/mesonbuild/meson/issues/13742

---

This is not a complete solution for #13740 and #13741, but would be a good starting point for fixing those afterwards.

### CODE REVIEW

File: mesonbuild/scripts/env2mfile.py [hunk 1]
```diff
@@ -180,7 +185,8 @@ def detect_cross_debianlike(options: T.Any) -> MachineInfo:
     host_arch = data['DEB_HOST_GNU_TYPE']
     host_os = data['DEB_HOST_ARCH_OS']
     host_subsystem = host_os
```
Comment (smcv) [inline mesonbuild/scripts/env2mfile.py:187]: Is this right? In practice `DEB_HOST_ARCH_OS` takes values `linux`, `kfreebsd` and `hurd`.

The only subsystems that Meson currently documents are the various flavours of Darwin (`macos`, `ios`, etc.), and if anything I would have expected the closest equivalent of that on Linux to be the libc implementation (the "GNU" in GNU/Linux). In Debian that's `DEB_HOST_ARCH_LIBC`, which in practice is `gnu` on all extant Debian architectures and ports, but ports that use `musl` or `uclibc` have been proposed.

Comment (smcv) [reply mesonbuild/scripts/env2mfile.py:187]: Or possibly the Meson subsystem is more like the `DEB_BUILD_ARCH_ABI`, which is normally `base`, but can be for example `eabihf` on ARM EABI hardfloat.

Or, if we don't actually know what "subsystem" would really mean on non-Darwin, maybe it would be better to not include it in `env2mfile` output at all.

Comment (dcbaker) [reply mesonbuild/scripts/env2mfile.py:187]: I don't think we should set them except on MacOS, since they're not documented in the table, or we should document them and hav env2mfile massage values to our expected ones. And the methods aren't documented in the function reference. Sweet.

tl;dr This was all a sort of "best we can agree on" with subsystem. I really wanted to have a clear way to differentiate Android/Linux, GNU/Linux, Debian/FreeBSD, FreeBSD/FreeBSD, GNU/OpenSolaris, etc. We just couldn't really get an agreement on what to call those things.

Comment (smcv) [reply mesonbuild/scripts/env2mfile.py:216]: I'm going to turn this thread into a separate issue report.

File: mesonbuild/scripts/env2mfile.py [hunk 2]
```diff
@@ -170,6 +200,9 @@
         cmd = ['dpkg-architecture', '-a' + options.debarch]
     output = subprocess.check_output(cmd, universal_newlines=True,
                                      stderr=subprocess.DEVNULL)
+    return dpkg_architecture_to_machine_info(output)
```
[the remark this answers is not in the document]
Comment (smcv) [reply mesonbuild/scripts/env2mfile.py]: fixed, thanks bot

File: unittests/internaltests.py [hunk 3]
```diff
@@ -23,6 +24,7 @@
 import mesonbuild.envconfig
 import mesonbuild.environment
 import mesonbuild.modules.gnome
+import mesonbuild.scripts.env2mfile
```
[the remark this answers is not in the document]
Comment (smcv) [reply unittests/internaltests.py:27]: fixed lint warning

File: mesonbuild/scripts/env2mfile.py [hunk 4]
```diff
@@ -48,20 +48,34 @@ def add_arguments(parser: 'argparse.ArgumentParser') -> None:
...
-
-        self.system: T.Optional[str] = None
-        self.subsystem: T.Optional[str] = None
-        self.kernel: T.Optional[str] = None
-        self.cpu: T.Optional[str] = None
-        self.cpu_family: T.Optional[str] = None
-        self.endian: T.Optional[str] = None
+    def __init__(
+        self,
+        compilers: T.Dict[str, T.List[str]] = {},
```
Comment (smcv) [inline mesonbuild/scripts/env2mfile.py:53]: pylint doesn't like this, but I'm using `dict()` to explicitly copy it, which avoids the pitfalls of mutable default arguments. I'll see whether there's a common solution to this elsewhere in the codebase, but in the meantime I'd appreciate review on the rest.

File: mesonbuild/scripts/env2mfile.py (hunk 1 above)
Comment (smcv) [reply mesonbuild/scripts/env2mfile.py:245]: #13743

File: mesonbuild/scripts/env2mfile.py [hunk 5]
```diff
@@ -63,6 +63,49 @@ def __init__(self) -> None:
...
+        if compile_args is not None:
+            self.compile_args.update(compile_args)
+
+        if link_args is not None:
+            self.link_args.update(link_args)
+
+        if cmake is not None:
+            self.cmake.update(cmake)
+
+        return self
```
Comment (dcbaker) [inline mesonbuild/scripts/env2mfile.py:107]: Is there really a need for a second initializer? This really seems like we could use a `dataclass` to simplify the default initializer to be more flexible:
```python
@dataclass
class MachineInfo:
    
    compilers: T.Dict[str, T.List[str]] = field(default_factory=dict)
    binaries: T.Dict[str, T.List[str]] = field(default_factory=dict)
    ....
    system: T.Optional[str] = None
```

File: mesonbuild/scripts/env2mfile.py [hunk 6]
```diff
@@ -178,13 +240,16 @@ def detect_cross_debianlike(options: T.Any) -> MachineInfo:
         k, v = line.split('=', 1)
         data[k] = v
     host_arch = data['DEB_HOST_GNU_TYPE']
-    host_os = data['DEB_HOST_ARCH_OS']
+    host_os = deb_os_map.get(data['DEB_HOST_ARCH_OS'],
+                             data['DEB_HOST_ARCH_OS'])
```
Comment (dcbaker) [inline mesonbuild/scripts/env2mfile.py:244]: It's a bit tangential, but this pattern of getting seems to be repeated a lot, maybe adding a helper like:
```python
def from_map(mapping: T.Mapping[str, str], name: str) -> str:
   return mapping.get(name, name)

host_os = from_map(deb_os_map, data['DEB_HOST_ARCH_OS'])
```

File: unittests/internaltests.py [hunk 7]
```diff
@@ -1715,3 +1717,475 @@ def test_option_key_from_string(self) -> None:
         for raw, expected in cases:
             with self.subTest(raw):
                 self.assertEqual(OptionKey.from_string(raw), expected)
+
+    def test_env2mfile_deb(self) -> None:
+        import argparse
```
Comment (dcbaker) [inline unittests/internaltests.py:1722]: This should be imported at the top level

File: unittests/internaltests.py [hunk 8]
```diff
@@ -1715,3 +1717,475 @@ def test_option_key_from_string(self) -> None:
...
+        MachineInfo = mesonbuild.scripts.env2mfile.MachineInfo
+        to_machine_info = mesonbuild.scripts.env2mfile.dpkg_architecture_to_machine_info
+
+        # For testing purposes, behave as though all cross-programs
+        # exist in /usr/bin
+        def locate_path(program: str) -> T.List[str]:
+            if os.path.isabs(program):
+                return [program]
+            else:
+                return ['/usr/bin/' + program]
```
Comment (dcbaker) [inline unittests/internaltests.py:1732]:
```suggestion
            return ['/usr/bin/' + program]
```

File: unittests/internaltests.py [hunk 9]
```diff
@@ -1715,3 +1717,475 @@ def test_option_key_from_string(self) -> None:
...
+            with self.subTest(title), \
+                    unittest.mock.patch.dict('os.environ', env, clear=True), \
+                    unittest.mock.patch('mesonbuild.scripts.env2mfile.locate_path') as mock_locate_path:
+                mock_locate_path.side_effect = locate_path
+                options = argparse.Namespace()
+                options.gccsuffix = gccsuffix
+                actual = to_machine_info(dpkg_arch, options)
+
+                if expected.system == 'TODO':
+                    print('TODO: %s: system() -> %s' % (title, actual.system))
```
Comment (dcbaker) [inline unittests/internaltests.py:2164]: We don't use the `%` formatters, either f-strings for str.format. In this case I think: `f'TODO: {title}: system() -> {actual.system}'` increases readability too

**Review by dcbaker — CHANGES_REQUESTED** [review]
Thanks for the very thorough testing of this, I've got a few comments below.

File: mesonbuild/scripts/env2mfile.py (hunk 5 above)
Comment (smcv) [reply mesonbuild/scripts/env2mfile.py:107]: I wasn't aware of dataclasses being available in Meson's minimum required Python. If this works, then yes it's probably the right way to achieve what I'm doing here.

File: mesonbuild/scripts/env2mfile.py (hunk 6 above)
Comment (smcv) [reply mesonbuild/scripts/env2mfile.py:244]: Maybe. My concern about this would be that it isn't necessarily obvious what the default is.

What's actually happening here is something like "identity mapping, with some special cases" so perhaps a slightly longer name than `from_map` could clarify that?

File: unittests/internaltests.py (hunk 9 above)
Comment (smcv) [reply unittests/internaltests.py:2164]: Clearly I spend too much time in ancient container environments (or ancient codebases) where f-strings are either nonexistent, or considered new and suspicious :-)

File: mesonbuild/scripts/env2mfile.py (hunk 5 above)
Comment (dcbaker) [reply mesonbuild/scripts/env2mfile.py:107]: We have dataclasses, Python 3.7 is the current minimum supported version IIRC

File: mesonbuild/scripts/env2mfile.py (hunk 6 above)
Comment (dcbaker) [reply mesonbuild/scripts/env2mfile.py:244]: Sure, I'm mostly concerned with the code repetition here, and the potential for typos that don't immediately register because we don't use the fallback in our test environment.

Comment (smcv) [reply mesonbuild/scripts/env2mfile.py:244]: I called it `replace_special_cases` which I think is clear enough.

**Review by dcbaker — APPROVED** [review]
Thanks, this looks good to me.
